src/index/VocabularyImpl.h

```diff
@@ -0,0 +1,347 @@
+  }
+  _words.resize(nofInternal);
+  _externalLiterals.buildFromVector(extVocab, fileName);
+  LOG(INFO) << "Done externalizing literals." << std::endl;
+}
+
+// _____________________________________________________________________________
+template <class S>
+bool Vocabulary<S>::isLiteral(const string& word) {
+  return word.size() > 0 && word[0] == '\"';
+}
+
+// _____________________________________________________________________________
+template <class S>
+template <bool isEntity>
+bool Vocabulary<S>::shouldBeExternalized(const string& word) const {
+  if constexpr (isEntity) {
+    return shouldEntityBeExternalized(word);
+  }
+
+  if (!isLiteral(word)) {
+    return shouldEntityBeExternalized(word);
+  } else {
+    if (word.size() > 100) {
+      return true;
+    }
+    string lang = getLanguage(word);
+    if (lang != "") {
+      return (lang != "en");  // && lang != "en_gb" && lang != "en_us" &&
+      // lang != "de" && lang != "es" && lang != "fr");
+    }
+  }
+  return false;
+}
+
+// partial instantiation, TODO: does itwork like this
+/*
+template <class S>
+template bool Vocabulary<S>::shouldBeExternalized<true>(
+    const string& word) const;
+
+template <class S>
+template bool Vocabulary<S>::shouldBeExternalized<false>(
+    const string& word) const;
+    */
+
+// ___________________________________________________________________
+template <class S>
+bool Vocabulary<S>::shouldEntityBeExternalized(const string& word) const {
+  for (const auto& p : _externalizedPrefixes) {
+    if (ad_utility::startsWith(word, p)) {
+      return true;
+    }
+  }
+  return false;
+}
+
+// _____________________________________________________________________________
+template <class S>
+string Vocabulary<S>::getLanguage(const string& literal) {
+  auto lioAt = literal.rfind('@');
+  if (lioAt != string::npos) {
+    auto lioQ = literal.rfind('\"');
+    if (lioQ != string::npos && lioQ < lioAt) {
+      return literal.substr(lioAt + 1);
+    }
+  }
+  return "";
+}
+
+// ____________________________________________________________________________
+template <class S>
+template <typename>
+string Vocabulary<S>::expandPrefix(const CompressedString& word) const {
+  assert(!word.empty());
+  auto idx = static_cast<unsigned char>(word[0]) - MIN_COMPRESSION_PREFIX;
+  if (idx < NUM_COMPRESSION_PREFIXES) {
+    return _prefixMap[idx] + word.toStringView().substr(1);
+  } else {
+    return string(word.toStringView().substr(1));
+  }
+}
+
+// _____________________________________________________________________________
+template <class S>
+template <typename>
+CompressedString Vocabulary<S>::compressPrefix(const string& word) const {
+  for (const auto& p : _prefixVec) {
+    if (ad_utility::startsWith(word, p._fulltext)) {
+      auto res = CompressedString::fromString(
+          p._prefix + std::string_view(word).substr(p._fulltext.size()));
+      LOG(DEBUG) << "compressed " << word << " to " << res.toString() << '\n';
```

Comment:
is that `toString()` needed? I'd think that `ostream` handles `string_view`?

Reply:
This is not a string view but a CompressedString. But I can teach ostream to do this, would probably be cleaner. But in this place I anyway want to throw out that LOG, because it is unreadable and much too verbose for every word. (This was for tracking a bug, that has been removed for some time now).